platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/ServiceScopeValidator.java

```diff
@@ -84,6 +135,30 @@ private String invalidScopeMessage(Class<?> serviceType, Class<? extends Scope>[
         );
     }
 
+    private String missingScopeMessage(Class<?> serviceType) {
+        return String.format(
+            "The service '%s' is registered in the '%s' scope but does not declare it. " +
+                "Add the '@ServiceScope()' annotation on '%s' with the '%s' scope.",
+            serviceType.getName(), scope.getSimpleName(), serviceType.getSimpleName(), scope.getSimpleName()
+        );
+    }
+
+    private String implementationWithMissingScopeMessage(Class<?> serviceType, Class<?> implementationType) {
+        return String.format(
+            "The service implementation '%s' is registered in the '%s' scope but does not declare it explicitly.\n" +
+                "The implementation appears to serve %s.\n" +
+                "Try the following:\n" +
+                "- If registered via an instance or implementation type then use an overload providing an explicit service type, e.g. 'ServiceRegistration.add(serviceType, implementationType)'\n" +
+                "- If registered via a creator-method in a service provider class then change the return type of the method to the service type\n" +
+                "- Alternatively, add the '@ServiceScope()' to the implementation type",
+            implementationType.getName(), scope.getSimpleName(), displayServiceTypes(serviceType)
+        );
+    }
```

Comment:
💭 A detailed message on what to do in the common case of a registering a service just via its implementation class. This happens mostly when a service is registered via `ServiceRegistration.add(serviceType)`, and in some other cases as well.

Reply:
In the future, it should also cover the case of when a single implementation serves multiple services.